cmd/activator/main.go

```diff
@@ -76,6 +76,18 @@ var (
 	reqChan  = make(chan activatorhandler.ReqEvent, requestCountingQueueLength)
 )
 
+type readinessHandler struct {
+	nextHandler http.Handler
+}
+
+func (rh *readinessHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.RequestURI == "/healthz" {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
```

Comment:
You shouldn't need to allocate a specific path for it. The kubernetes probes set a specific header that you can check for. See the queue-proxy for an example on that.

I guess we could then factor this handler to a common place and use it both in the queue-proxy and in the activator 🙂 .

Reply:
See https://github.com/knative/serving/blob/master/cmd/queue/main.go#L161-L165

Reply:
I cant come up with a good argument one way or the other (explicit path for implicit useragent check), but I really like the idea of factoring this in to a common place - Its a good idea for us to at least perform the same type of check everywhere we do.

Reply:
It was not obvious to me how the `isProbe` function, which checks the user-agent header, is used because I do not see it declared as readiness probe. @greghaynes told me that it may be there for historical reasons. It was introduced in [this PR](https://github.com/knative/serving/pull/670). 

I did a small test in which I added a log statement to the `isProbe` function to see if the kube-probe user-agent header is ever present in the request. I never saw it.

I think that the `/healthz` endpoint is reasonable since it seems to be the typical approach in the [docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/).

I went ahead and moved the readiness handler into a different package for reuse with queue-proxy; An http readiness probe is declared for queue-proxy at the `/health` endpoint. [That code](https://github.com/knative/serving/blob/master/cmd/queue/main.go#L198) looks like it just wraps a TCP probe to the user-container. I need to look it over a bit more and then make sure the same thing is being done for both activator and queue-proxy, and then I will refactor the common code out to eliminate duplication.

Reply:
To fill in the missing pieces of history and why things are the way they are:

The `isProbe` function in the queue-proxy is used to catch a probe that goes to the **user-container**. That sounds counter-intuitive at first but hear me out: Istio rewriting affects the traffic in the container itself as well. A HTTP probe on the user-container (as defined in the `revisionTemplate`) is rewritten to go through the istio-proxy to the queue-proxy and finally into the user-container. In the queue-proxy, we do not want to count these requests as actual requests since that would cause us to never scale down to zero, hence we filter them out using that header that Kubernetes always sets.

I do realize though that we need slightly different behavior which will make reuse limited. In the queue-proxy we proxy requests with those headers happily but exclude them from concurrency measurements. In the activator though, we need to *respond* to those requests vs. passing them through to the revision.

I think we can put the `isK8sProbe` function in a common place, but we'll still need 2 different handlers to accomplish that difference in behavior.

That being said, I think we absolutely should use that header vs. a common path like `/healthz`. Keep in mind that this will prevent all user apps from using `/healthz` as a path on their app since the activator will intercept those calls.

Reply:
Ah thanks for the info @markusthoemmes, that makes a lot of sense. I agree with you that we dont want to mask a path in the activator from being able to be used for a user's app. What do you think about doing a pattern similar to queue-proxy where we have a dedicated port for things like this (adminServer)? Then we are truly not masking *anything* from a user's app and are a bit more explicit/standard in how health checks are performed.